tests/common/glcubetexturelayerrenderer.cpp

```diff
@@ -0,0 +1,622 @@
+};
+
+class StreamTextureImpl
+{
+public:
+   ~StreamTextureImpl ()
+  {
+    glDeleteTextures (1, &gl_tex_);
+    gl_->eglDestroyImageKHR (gl_->display, image_);
+    close (fd_);
+    gbm_bo_destroy (bo_);
+  }
+
+  void *Map ()
+  {
+    assert (addr_ == nullptr);
+    size_t size = dimension_.stride * dimension_.height;
+    addr_ =
+      mmap (nullptr, size, (PROT_READ | PROT_WRITE), MAP_SHARED, fd_, 0);
+    if (addr_ == MAP_FAILED)
+      return nullptr;
+    return addr_;
+  }
+
+  void Unmap ()
+  {
+    assert (addr_ != nullptr);
+    size_t size = dimension_.stride * dimension_.height;
+    munmap (addr_, size);
+    addr_ = nullptr;
+  }
+
+  GLuint GetTextureID () const
+  {
+    return gl_tex_;
+  }
+
+  Dimension GetDimension () const
+  {
+    return dimension_;
+  }
+
+  bool Initialize (struct gbm_device *gbm)
+  {
+    bo_ = gbm_bo_create (gbm, dimension_.width, dimension_.height,
+			 GBM_FORMAT_ARGB8888, GBM_BO_USE_LINEAR);
+    if (!bo_)
+      {
+	printf ("failed to create a gbm buffer.\n");
+	return false;
+      }
+
+    fd_ = gbm_bo_get_fd (bo_);
+    if (fd_ < 0)
+      {
+	printf ("failed to get fb for bo: %d\n", fd_);
+	return false;
+      }
+
+    dimension_.stride = gbm_bo_get_stride (bo_);
+    EGLint offset = 0;
+    const EGLint khr_image_attrs[] = { EGL_DMA_BUF_PLANE0_FD_EXT,
+      fd_,
+      EGL_WIDTH,
+      dimension_.width,
+      EGL_HEIGHT,
+      dimension_.height,
+      EGL_LINUX_DRM_FOURCC_EXT,
+      GBM_FORMAT_ARGB8888,
+      EGL_DMA_BUF_PLANE0_PITCH_EXT,
+      dimension_.stride,
+      EGL_DMA_BUF_PLANE0_OFFSET_EXT,
+      offset,
+      EGL_NONE
+    };
+
+    image_ = gl_->eglCreateImageKHR(gl_->display, EGL_NO_CONTEXT,
+                                      EGL_LINUX_DMA_BUF_EXT,
+                                      nullptr, khr_image_attrs);
+
+    if (image_ == EGL_NO_IMAGE_KHR)
+      {
+	printf ("failed to make image from buffer object: %x\n", glGetError());
+	return false;
+      }
+
+    glGenTextures (1, &gl_tex_);
+    glBindTexture (GL_TEXTURE_2D, gl_tex_);
+    gl_->glEGLImageTargetTexture2DOES (GL_TEXTURE_2D, image_);
+    glTexParameteri (GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
+    glTexParameteri (GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
+    glBindTexture (GL_TEXTURE_2D, 0);
+    return true;
+  }
+
+  StreamTextureImpl (glContext *gl, size_t width,
+		       size_t height)
+  {
+    gl_ = gl;
+    dimension_.width = width;
+    dimension_.height = height;
+  }
+
+private:
+  glContext *gl_;
+  struct gbm_bo *bo_ = nullptr;
+  int fd_ = -1;
+  EGLImageKHR image_ = nullptr;
+  GLuint gl_tex_ = 0;
+  Dimension dimension_;
+  void *addr_ = nullptr;
+};
+
+GLCubeTextureLayerRenderer::GLCubeTextureLayerRenderer (struct gbm_device *dev):
+GLLayerRenderer (dev)
+{
+}
+
+GLCubeTextureLayerRenderer::~GLCubeTextureLayerRenderer ()
+{
+}
+
+bool
+  GLCubeTextureLayerRenderer::Init (uint32_t width, uint32_t height,
+			       uint32_t format, glContext * gl,
```

Comment:
Can we make this common helper class and use it here and https://github.com/01org/IA-Hardware-Composer/blob/master/tests/common/glcubelayerrenderer.cpp#L28